File: Python/CanBattle.py
from Karakter import Karakter

class CanBattle(Karakter):
    """Karakter yang bisa ikut pertarungan."""
    def __init__(self, name: str, gender: str, race: str, affiliation,
                 maxAP: int, maxHP: int, maxM: int, charClass: str, **kwargs):
        super().__init__(name, gender, race, affiliation, **kwargs) # Meneruskan ke Karakter
        self.maxActionPoint = maxAP
        self.maxHealth = maxHP
        self.maxMana = maxM
        self.charClass = charClass

    def display(self):
        super().display()
        print(f"  Kelas: {self.charClass}")
        print(f"  Max HP: {self.maxHealth}")
        print(f"  Max Mana: {self.maxMana}")
        print(f"  Max AP: {self.maxActionPoint}")

File: C++/CanBattle.cpp
#pragma once
#include "Karakter.cpp"

class CanBattle : public virtual Karakter { // <-- VIRTUAL INHERITANCE
protected:
    int maxActionPoint;
    int maxHealth;
    int maxMana;
    std::string charClass; // class

public:
    CanBattle(std::string name, std::string gender, std::string race, Factions* affiliation,
              int maxAP, int maxHP, int maxM, std::string charClass)
        // Karakter(name, gender, race, affiliation) tidak dieksekusi oleh CanBattle jika diwariskan secara virtual
        // kecuali CanBattle adalah kelas yang paling diturunkan.
        : Karakter(name, gender, race, affiliation), // Tetap ada, tapi hanya dipanggil jika CanBattle bukan virtual base.
          maxActionPoint(maxAP), maxHealth(maxHP), maxMana(maxM), charClass(charClass) {}

    // Getter
    int getMaxActionPoint() const { return maxActionPoint; }
    int getMaxHealth() const { return maxHealth; }
    int getMaxMana() const { return maxMana; }
    std::string getCharClass() const { return charClass; }

    // Setter
    void setMaxActionPoint(int ap) { maxActionPoint = ap; }
    void setMaxHealth(int hp) { maxHealth = hp; }
    void setMaxMana(int m) { maxMana = m; }
    void setCharClass(std::string c) { charClass = c; }

    // Override Display
    void display() const override {
        Karakter::display();
        std::cout << "  Kelas: " << charClass << std::endl;
        std::cout << "  Max HP: " << maxHealth << std::endl;
        std::cout << "  Max Mana: " << maxMana << std::endl;
        std::cout << "  Max AP: " << maxActionPoint << std::endl;
    }
};